library/modules/Military.cpp

```diff
@@ -289,3 +295,116 @@ void Military::updateRoomAssignments(int32_t squad_id, int32_t civzone_id, df::s
+{
+    int32_t start_year = -1;
+    for (size_t i = 0; i < hf->entity_links.size(); i++)
+    {
+        auto link = strict_virtual_cast<df::histfig_entity_link_squadst>(hf->entity_links[i]);
+        if (link == nullptr || link->squad_id != squad->id)
+            continue;
+
+        hf->entity_links.erase(hf->entity_links.begin() + i);
+        start_year = link->start_year;
+
+        delete link;
+        break;
+    }
+
+    if (start_year == -1)
+        return false;
+
+    auto former_squad = df::allocate<df::histfig_entity_link_former_squadst>();
+    former_squad->squad_id = squad->id;
+    former_squad->entity_id = squad->entity_id;
+    former_squad->start_year = start_year;
+    former_squad->end_year = *df::global::cur_year;
+    former_squad->link_strength = 100;
+
+    hf->entity_links.push_back(former_squad);
+    return true;
+}
+
+static bool remove_officer_entity_link(df::historical_figure* hf, df::squad* squad)
+{
+    std::vector<Units::NoblePosition> nps;
+    if (!Units::getNoblePositions(&nps, hf))
+        return false;
+
+    int32_t assignment_id = -1;
+    for (auto& np : nps)
+    {
+        if (np.entity->id != squad->entity_id || np.assignment->squad_id != squad->id)
+            continue;
+
+        np.assignment->histfig = -1;
+        np.assignment->histfig2 = -1;
+
+        assignment_id = np.assignment->id;
+        break;
+    }
+
+    if (assignment_id == -1)
+        return false;
+
+    int32_t start_year = -1;
+    for (size_t i = 0; i < hf->entity_links.size(); i++)
+    {
+        auto link = strict_virtual_cast<df::histfig_entity_link_positionst>(hf->entity_links[i]);
+        if (link == nullptr)
+            continue;
+        if (link->assignment_id != assignment_id && link->entity_id != squad->entity_id)
+            continue;
+
+        hf->entity_links.erase(hf->entity_links.begin() + i);
+        start_year = link->start_year;
+
+        delete link;
+        break;
+    }
+
+    if (start_year == -1)
+        return false;
+
+    auto former_pos = df::allocate<df::histfig_entity_link_former_positionst>();
+    former_pos->assignment_id = assignment_id;
+    former_pos->entity_id = squad->entity_id;
+    former_pos->start_year = start_year;
+    former_pos->end_year = *df::global::cur_year;
+    former_pos->link_strength = 100;
+
+    hf->entity_links.push_back(former_pos);
+    return true;
+}
+
+bool Military::removeFromSquad(int32_t unit_id)
+{
+    df::unit *unit = df::unit::find(unit_id);
+    if (unit == nullptr)
+        return false;
+    if (unit->military.squad_id == -1 || unit->military.squad_position == -1)
+        return false;
+
+    int32_t squad_id = unit->military.squad_id;
+    df::squad* squad = df::squad::find(squad_id);
+    if (squad == nullptr)
+        return false;
+
+    // remove from squad information
+    int32_t squad_pos = unit->military.squad_position;
+    df::squad_position* pos = squad->positions.at(squad_pos);
```

Comment:
`at` can crash if given bad data. use `vector_get`, which is a safe wrapper. Be sure to test the return for null